src/gc-arena/src/context.rs

```diff
@@ -174,13 +164,18 @@ impl Context {
     //
     // In order for this to be safe, at the time of call no `Gc` pointers can be live that are not
     // reachable from the given root object.
-    pub(crate) unsafe fn do_collection(&mut self, work: f64) -> f64 {
+    pub(crate) unsafe fn do_collection<R: Collect>(&self, root: &R, work: f64) -> f64 {
         let mut work_done = 0.0;
         let cc = CollectionContext { context: self };
 
         while work > work_done {
             match self.phase.get() {
                 Phase::Propagate => {
+                    if self.root_needs_trace.get() {
+                        root.trace(cc);
+                        self.root_needs_trace.set(false);
+                    }
+
```

Comment:
I think it would be more efficient to only trace the root at the end of the `Propagate` phase (when the gray queues are empty), so that multiple consecutive calls to `root_barrier` can be coalesced.

Reply:
Yeah, you're right..